From patchwork Tue Sep  6 19:48:57 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968095
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id DF70AC54EE9
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:50:09 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id 98C0880015; Tue,  6 Sep 2022 15:49:53 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 8C71B80007; Tue,  6 Sep 2022 15:49:53 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 655BC80015; Tue,  6 Sep 2022 15:49:53 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0013.hostedemail.com
 [216.40.44.13])
	by kanga.kvack.org (Postfix) with ESMTP id 41DA280007
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:49:53 -0400 (EDT)
Received: from smtpin11.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay01.hostedemail.com (Postfix) with ESMTP id 1741A1C63CF
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:53 +0000 (UTC)
X-FDA: 79882701066.11.2F6DB6C
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf17.hostedemail.com (Postfix) with ESMTP id A5B5640070
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:52 +0000 (UTC)
Received: from pps.filterd (m0246617.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286IdYZj029954;
	Tue, 6 Sep 2022 19:49:45 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=Ngf70k76kRpRObHjvGvMUliwq0ETt1cTVrTppxrchEY=;
 b=NEs4xz8YVuDLXaq/CCVgqa+I/wwzHu8Aj5OabE4t0LOzb/GRmGlv5M4Kwo9rHvtbv3QI
 ZHq5Wo3cLNtj/bbQQlGpQPEsga+KuwLXtdlkPBuTIUNfr9upz4iPTyj4sDsVr0shbj3J
 jyYKKv2eBt0LplDOGTivj160QgDuKjkPcxKd7IlIOFz+4hdTg3/+ptKlzef78c2a3b6B
 gPxpWYRX+Jx1TxmDi4+QhJ8KdiY5naxs1DGp+78PhISZO76JsL2I6o5lLLJTR+ChaC91
 JpAAUDfsheMX5ruhlJbFAzF/QSTuMSJ8S9dE67XUj4Elx3A6sahCsG9cJS2O2pdLMkfe /w==
Received: from iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com
 (iadpaimrmta03.appoci.oracle.com [130.35.103.27])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbyftpvd9-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:45 +0000
Received: from pps.filterd
 (iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com [127.0.0.1])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286Je5WX039711;
	Tue, 6 Sep 2022 19:49:44 GMT
Received: from nam10-dm6-obe.outbound.protection.outlook.com
 (mail-dm6nam10lp2108.outbound.protection.outlook.com [104.47.58.108])
	by iadpaimrmta03.imrmtpd1.prodappiadaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc9m7j9-10
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:49:44 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=ntfklo+N13TPQOG2VIUwNP7V1050QUnRTlEWmifkuN5W1WUroNwa1A6pxZSReW9JagDV1lTMUE0IglJT39mpWvK5agX7tSe8ms7CNtadru6euKREvgzi0rsF45vb7D0cZ7DVmIQYQb/t81r82Q+IkGVcSlIIvunsPu1K75Mb05vZbU8I9XVzfdSfDLHcJiXLMdXTqSY/pEkVOQ2Sv8mAW2HqBasjuRexDQDfUy3fqwp1KmhUt5kWkH0FZa3Qxr8qX7q5Y9TwiQbPMT82H2RVi3jve8gvomAUkJYs2NePgblcVm1BmwISLHkJJkHMQR5bF0cg76d1kbf/ptZ9F16Ayg==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=Ngf70k76kRpRObHjvGvMUliwq0ETt1cTVrTppxrchEY=;
 b=CIv7Z5Drgg+3pmjfwMgiRNeKe2/hizRTZSw1Tt+r1+o2p6BNabGApsuwQ1+notgPiu3DR0WAaJVaq3+kH9nIlbFWwU6UGySLTeCt32Z8+nw4DQbnv1dSLSh69nrGiWp3S1W0Aoho4iO2mWsEOde3iLrt4V/v++/fU2NpJhD2+yOhQcGb0PbR+b3ZLf6n5bjMEgjDRahCRQ267zSB493a2EFePa+oRs/y1sQQzjwxVGa3jbUCMaktUFzpMyYGyeF/2T/Zm9jcc8VTrAm034OJjCLRBnhwqA66oF4sPYGwwRQYbvxwhbJBDeezQLQraC+kirSdTd2/EeFt6YnugSaIWQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Ngf70k76kRpRObHjvGvMUliwq0ETt1cTVrTppxrchEY=;
 b=kwh3yJQGLoeIsG/JzYdPtU9wJWiR0/aa8uA93e6wHFYWOuYwu2fqtgA2RGwV3LECXmkwxLOqDZpTKpeN7eEKBJy6jOj4qRHaXuIq5IOrhDhuU7Pqkspobj9ZxY9Sx2XX9pEHCNUgztl4IxT2IyY27g0S489rS1HKOCqQ4iMfIxc=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by SJ0PR10MB5598.namprd10.prod.outlook.com (2603:10b6:a03:3d9::18) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.10; Tue, 6 Sep
 2022 19:49:41 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:49:41 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: "Matthew Wilcox (Oracle)" <willy@infradead.org>,
        Liam Howlett
	<liam.howlett@oracle.com>
Subject: [PATCH v14 43/70] fs/proc/task_mmu: stop using linked list and
 highest_vm_end
Thread-Topic: [PATCH v14 43/70] fs/proc/task_mmu: stop using linked list and
 highest_vm_end
Thread-Index: AQHYwimzZL/z5I7PC02dI3D7mIyPWA==
Date: Tue, 6 Sep 2022 19:48:57 +0000
Message-ID: <20220906194824.2110408-44-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 9f4c65ee-516b-4447-1b57-08da9040f08f
x-ms-traffictypediagnostic: SJ0PR10MB5598:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 SgHWynPjcDiLibgCQQi9D8qSCNb5Fpve1T70kqX5XuMSnhjxswaWQU7GRp3PmeZOPgaxKxxtwb8VcxW4uz3M1KgXJ/wzgxtGgSmi0NdeREJZdQvM10WHy5tNvXTE0Sd0T/KXU/KhdiC7Y12EXip/0adKKZrs9X4AlsjahvUgIAChTfrReIfdI+Vb/BAMQgWrbhtiYQR+7ebj+YkfDQqpodKAuA2RJ6RJTV2NCcY6kFAmYK9DE6s/RPMAFMM4MWZjZsfxWqfQWPJYZDFqHZu8aoTiesKQ09q4aWPhOoiElSCcrHwffXYY3Jz4yXl5OazjR+QcEjUvMOtl7wMe353SsM+mS2b2gtJm+B02m1QyspA11N7RxlQt1IfqNVkPfW18q4DzbSSSSyNLIyG8dGujhMnsa/ODFZhl+5vOHQq9PIMJtTfwfq+EJ1jQMBptcpI6XCgbEeW93liEgS+fpQRs58oHGpxDxBzDPvOShu9G+D2VLOekYbagYmOxh4X/2JXDfF+ze9DLctOVUbI35jiaSMRXt/ckI6MF6yPr7FViHv4dWdsDnX6xmu5XkOYsXO/aymG2MUtnvrjc5hUYcPCELACDs5kf9Uh1LKpYA3OtFD2x/CLZb/GE17SAK6ApXZan/n2fkm0Jn5gH8sCW5prTwqoTscXwDpxhnaaTeYav5UJeeasfXNPCKIK85D/QEOIZOEexQnAfHtAKoadnwMmAiMXFCv0f0iWdNenp8j5WysuIEjcJFipLBDZK2v8wGPHM0FxF9QebuEde4oS3tyWkkQ==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(346002)(39860400002)(396003)(136003)(376002)(366004)(64756008)(66556008)(41300700001)(91956017)(76116006)(66476007)(66446008)(66946007)(71200400001)(54906003)(110136005)(36756003)(316002)(6512007)(44832011)(8676002)(122000001)(2906002)(6486002)(4326008)(186003)(8936002)(5660300002)(478600001)(38070700005)(38100700002)(2616005)(86362001)(26005)(1076003)(107886003)(6666004)(6506007)(83380400001);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?gvr4z+DilLgC1DExh6HbtIx?=
	=?iso-8859-1?q?/Fc71ZHa1Udi5SyIiQ9LBpFtHpCzMGrtFPzZRgfoyzThXvbRT04I+E2jEa50?=
	=?iso-8859-1?q?I7u8nRRAZ1pWzF8I9grIODg2qqqwhF2RjorUTBwXkZO1wJOXD9xxEE417ELk?=
	=?iso-8859-1?q?WepttAc0QCLrLK+HpUnyjbHBwhUbRxC1quEqo/k1bBi8sdgkL627W80I3WsB?=
	=?iso-8859-1?q?vxLOPW4MojKwQqklojrvznSfh5K1be1Nec15yYYwBvhefV9Yq+97LD8W3gv5?=
	=?iso-8859-1?q?+7fDsmrGLuoUKYPjWU4xP0AS4XKSpKQTawV+sZn6wYGmEIKNys6uW5RrCAQu?=
	=?iso-8859-1?q?i+8sDhprkBZ+xB9Kixgs68exQOl+wSb4WpZVPOr8Pmk8kuXfLc14QodyQrbJ?=
	=?iso-8859-1?q?e41flNCHvytMjTQC9aLZR7rDHL9Ua29KdrvBBuxR5MWnTPUSYl107OKmFn3m?=
	=?iso-8859-1?q?MHuokZOP0STZoxJeSxvpwoayfN9B4aSwdfckMankJOtvh1ToiLYQyq7h5y24?=
	=?iso-8859-1?q?Y1w7sOisDbkzAJiWjXLMn5uo/B6Wm4THvSsaFvt60LYokAunTQlA9Gn+cgf+?=
	=?iso-8859-1?q?AxMCd9Tk88py2Hlez53fMYHO8mwJEh4Ygo66WjRbYRxsVhmOKbc4ShuEyLU6?=
	=?iso-8859-1?q?d605WWU9TVcm7JE7E07vVAl3hnbCPZqUzSV7IQgLZWAsSrM8FOMLH+rYpG4F?=
	=?iso-8859-1?q?/ryHcOFOlaikENxEWq0w4EImSwRtg0YITaD7kr7NKKWo0g6HwDGVedsUENho?=
	=?iso-8859-1?q?2qTBkxXqhJFFyEe+hS795kfnArQK75xtJnRthOgwSWG94iXmFgxXRTCb/UDB?=
	=?iso-8859-1?q?hTdnWPtud69/StfU6JciO+ycuK1xZ8YpxOg9cOSw+FM0/PdftGjh3AVllT3B?=
	=?iso-8859-1?q?nUARx/l8x4sIt/ElTsIMNyKp6VKqakIl2qmEK5JUtQbtaLWYI9T5KqEvECGW?=
	=?iso-8859-1?q?ahAn4+f5NpES/A6IIJILfkDsQ1BkQyl2QShlDXKidNv6mFu1MelzKn7D9E/z?=
	=?iso-8859-1?q?NoVPpfcS828hBqB7M3B3MbGFWiGuWn60ItGsBdq/5c/Xfc5rq9/0RSbQDb/+?=
	=?iso-8859-1?q?KLinG/X9e6ssHSUoiwAF2f6G6mWi2F9DQNzRRlpU6X1u0t5a7xyb3VntpOjr?=
	=?iso-8859-1?q?/hlbAIfDaCdycTUi3i6uEV5rPHOSmIJGp5VVI6UARnrMVpPPoQmTntKaaI7X?=
	=?iso-8859-1?q?1bjRzWvlkbnvpkaXJOE3gqpiS2d2J8Dm0ykEPZW8onClgM0yMWSBXuXuTowh?=
	=?iso-8859-1?q?tx+kPyYRQJtfnengAdrprRIRkHDSX/kUyQZtgJ9JFOR2A42pfCm0EI0ERGuh?=
	=?iso-8859-1?q?I0URVPcVWUJ1yhG+7l/GUAmEPO+KcIpcx/Yka+FIsgsO9wKfLrq1eES0wlVN?=
	=?iso-8859-1?q?xxDWlhXj3tztk8xfDt6LpzFNhwnVGFrG3RvwgXlJ2MmU6DYQEQKRazePNWih?=
	=?iso-8859-1?q?XGyIrPaz1lVUCAnWtYRuQGSw1D+B4CHXDBSCo3PYNuvC016pkAXR59NPA0bI?=
	=?iso-8859-1?q?kVNcuMKADsOVLnD2IGe6hl/fdvwT/KvL0vNDNmnMX+Xmik+MP/jcqq08uNkR?=
	=?iso-8859-1?q?XrTHQpZYvtQK4P/UNfVvw+W5aFMnBK3+Kn77V05Ecd3L8sJ09dGLrz1EKgf9?=
	=?iso-8859-1?q?b2PkuaZmRHyoI4geGXVV9EdFkaSIJox93gnB5fw=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 9f4c65ee-516b-4447-1b57-08da9040f08f
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:48:57.4593
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 PPDIpTwdArqpR9vzJJ3JLt4KTboIQfooCZ0CbYmjnVl87sELxhgbRoKffmtSzgMBvbcEgn3o/Rc6OX/SNAIEBg==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SJ0PR10MB5598
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 mlxlogscore=999
 phishscore=0 mlxscore=0 spamscore=0 malwarescore=0 suspectscore=0
 adultscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2207270000 definitions=main-2209060091
X-Proofpoint-ORIG-GUID: 4IZ5CGWr7eUUf3GuW5HW4L4Dz3xbi3O2
X-Proofpoint-GUID: 4IZ5CGWr7eUUf3GuW5HW4L4Dz3xbi3O2
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493792;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=Ngf70k76kRpRObHjvGvMUliwq0ETt1cTVrTppxrchEY=;
	b=1WmjAxNGEKsgwT0Ibfj7N2Z0OKB/MLI/9AnaE4MV8Ras93har0NFSrbVj1gaxkLOJWforl
	/DMA9CfgykE1F4Ov02L6iDQI0gzQbWXKUjE7BXCaUNTXpRjytUu7z3miGp5mNnso7P8ECt
	/fBDi0OWKBhP0RlgaXhMfNN+qQ06GlQ=
ARC-Authentication-Results: i=2;
	imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=NEs4xz8Y;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=kwh3yJQG;
	spf=pass (imf17.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1");
	dmarc=pass (policy=none) header.from=oracle.com
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493792; a=rsa-sha256;
	cv=pass;
	b=ySnmhsn0tLxCRigUzw1fU38a/Xm5ej2FfcLFEJSq0eB/S0WelyI652Tiren9jtZiaIVJ4Y
	EiOOObSOFclM7/MEHZp8E1o+m5mysTw8BaCenkFxc2qV+xFsrNWrF2Kxtk/0AM5t6y41oo
	t7gV3XsYzsy35nXMCzCvNTUF+R8VuUw=
X-Stat-Signature: hp59ois73r9bidcbihot613s4h1t556k
X-Rspam-User: 
X-Rspamd-Queue-Id: A5B5640070
Authentication-Results: imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=NEs4xz8Y;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=kwh3yJQG;
	spf=pass (imf17.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1");
	dmarc=pass (policy=none) header.from=oracle.com
X-Rspamd-Server: rspam01
X-HE-Tag: 1662493792-864205
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Matthew Wilcox (Oracle)" <willy@infradead.org>

Remove references to mm_struct linked list and highest_vm_end for when
they are removed

Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
---
 fs/proc/internal.h |  2 +-
 fs/proc/task_mmu.c | 73 ++++++++++++++++++++++++++--------------------
 2 files changed, 42 insertions(+), 33 deletions(-)

diff --git a/fs/proc/internal.h b/fs/proc/internal.h
index 06a80f78433d..f03000764ce5 100644
--- a/fs/proc/internal.h
+++ b/fs/proc/internal.h
@@ -285,7 +285,7 @@ struct proc_maps_private {
 	struct task_struct *task;
 	struct mm_struct *mm;
 #ifdef CONFIG_MMU
-	struct vm_area_struct *tail_vma;
+	struct vma_iterator iter;
 #endif
 #ifdef CONFIG_NUMA
 	struct mempolicy *task_mempolicy;
diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index 9f70bc1c2766..8b4f3073f8f5 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -123,12 +123,26 @@ static void release_task_mempolicy(struct proc_maps_private *priv)
 }
 #endif
 
+static struct vm_area_struct *proc_get_vma(struct proc_maps_private *priv,
+						loff_t *ppos)
+{
+	struct vm_area_struct *vma = vma_next(&priv->iter);
+
+	if (vma) {
+		*ppos = vma->vm_start;
+	} else {
+		*ppos = -2UL;
+		vma = get_gate_vma(priv->mm);
+	}
+
+	return vma;
+}
+
 static void *m_start(struct seq_file *m, loff_t *ppos)
 {
 	struct proc_maps_private *priv = m->private;
 	unsigned long last_addr = *ppos;
 	struct mm_struct *mm;
-	struct vm_area_struct *vma;
 
 	/* See m_next(). Zero at the start or after lseek. */
 	if (last_addr == -1UL)
@@ -152,31 +166,21 @@ static void *m_start(struct seq_file *m, loff_t *ppos)
 		return ERR_PTR(-EINTR);
 	}
 
+	vma_iter_init(&priv->iter, mm, last_addr);
 	hold_task_mempolicy(priv);
-	priv->tail_vma = get_gate_vma(mm);
-
-	vma = find_vma(mm, last_addr);
-	if (vma)
-		return vma;
+	if (last_addr == -2UL)
+		return get_gate_vma(mm);
 
-	return priv->tail_vma;
+	return proc_get_vma(priv, ppos);
 }
 
 static void *m_next(struct seq_file *m, void *v, loff_t *ppos)
 {
-	struct proc_maps_private *priv = m->private;
-	struct vm_area_struct *next, *vma = v;
-
-	if (vma == priv->tail_vma)
-		next = NULL;
-	else if (vma->vm_next)
-		next = vma->vm_next;
-	else
-		next = priv->tail_vma;
-
-	*ppos = next ? next->vm_start : -1UL;
-
-	return next;
+	if (*ppos == -2UL) {
+		*ppos = -1UL;
+		return NULL;
+	}
+	return proc_get_vma(m->private, ppos);
 }
 
 static void m_stop(struct seq_file *m, void *v)
@@ -876,16 +880,16 @@ static int show_smaps_rollup(struct seq_file *m, void *v)
 {
 	struct proc_maps_private *priv = m->private;
 	struct mem_size_stats mss;
-	struct mm_struct *mm;
+	struct mm_struct *mm = priv->mm;
 	struct vm_area_struct *vma;
-	unsigned long last_vma_end = 0;
+	unsigned long vma_start = 0, last_vma_end = 0;
 	int ret = 0;
+	MA_STATE(mas, &mm->mm_mt, 0, 0);
 
 	priv->task = get_proc_task(priv->inode);
 	if (!priv->task)
 		return -ESRCH;
 
-	mm = priv->mm;
 	if (!mm || !mmget_not_zero(mm)) {
 		ret = -ESRCH;
 		goto out_put_task;
@@ -898,8 +902,13 @@ static int show_smaps_rollup(struct seq_file *m, void *v)
 		goto out_put_mm;
 
 	hold_task_mempolicy(priv);
+	vma = mas_find(&mas, 0);
+
+	if (unlikely(!vma))
+		goto empty_set;
 
-	for (vma = priv->mm->mmap; vma;) {
+	vma_start = vma->vm_start;
+	do {
 		smap_gather_stats(vma, &mss, 0);
 		last_vma_end = vma->vm_end;
 
@@ -908,6 +917,7 @@ static int show_smaps_rollup(struct seq_file *m, void *v)
 		 * access it for write request.
 		 */
 		if (mmap_lock_is_contended(mm)) {
+			mas_pause(&mas);
 			mmap_read_unlock(mm);
 			ret = mmap_read_lock_killable(mm);
 			if (ret) {
@@ -951,7 +961,7 @@ static int show_smaps_rollup(struct seq_file *m, void *v)
 			 *    contains last_vma_end.
 			 *    Iterate VMA' from last_vma_end.
 			 */
-			vma = find_vma(mm, last_vma_end - 1);
+			vma = mas_find(&mas, ULONG_MAX);
 			/* Case 3 above */
 			if (!vma)
 				break;
@@ -965,11 +975,10 @@ static int show_smaps_rollup(struct seq_file *m, void *v)
 				smap_gather_stats(vma, &mss, last_vma_end);
 		}
 		/* Case 2 above */
-		vma = vma->vm_next;
-	}
+	} while ((vma = mas_find(&mas, ULONG_MAX)) != NULL);
 
-	show_vma_header_prefix(m, priv->mm->mmap->vm_start,
-			       last_vma_end, 0, 0, 0, 0);
+empty_set:
+	show_vma_header_prefix(m, vma_start, last_vma_end, 0, 0, 0, 0);
 	seq_pad(m, ' ');
 	seq_puts(m, "[rollup]\n");
 
@@ -1262,6 +1271,7 @@ static ssize_t clear_refs_write(struct file *file, const char __user *buf,
 		return -ESRCH;
 	mm = get_task_mm(task);
 	if (mm) {
+		MA_STATE(mas, &mm->mm_mt, 0, 0);
 		struct mmu_notifier_range range;
 		struct clear_refs_private cp = {
 			.type = type,
@@ -1307,8 +1317,8 @@ static ssize_t clear_refs_write(struct file *file, const char __user *buf,
 
 		if (start || end != -1UL)
 		{
-			start = min(start, mm->highest_vm_end) & PAGE_MASK;
-			end = min(end, mm->highest_vm_end) & PAGE_MASK;
+			start &= PAGE_MASK;
+			end &= PAGE_MASK;
 
 			if (start >= end)
 			{
@@ -1347,7 +1357,7 @@ static ssize_t clear_refs_write(struct file *file, const char __user *buf,
 				goto out_mm;
 			}
 			if (!clear_range)
-				for (vma = mm->mmap; vma; vma = vma->vm_next) {
+				mas_for_each(&mas, vma, ULONG_MAX) {
 				    if (!(vma->vm_flags & VM_SOFTDIRTY))
 					    continue;
 				    mmap_read_unlock(mm);
@@ -1355,7 +1365,7 @@ static ssize_t clear_refs_write(struct file *file, const char __user *buf,
 					    count = -EINTR;
 					    goto out_mm;
 				    }
-				    for (vma = mm->mmap; vma; vma = vma->vm_next) {
+				    mas_for_each(&mas, vma, ULONG_MAX) {
 					    vma->vm_flags &= ~VM_SOFTDIRTY;
 					    vma_set_page_prot(vma);
 				    }
@@ -1374,8 +1384,7 @@ static ssize_t clear_refs_write(struct file *file, const char __user *buf,
 				goto out_mm;
 			}
 		}
-		walk_page_range(mm, start, end == -1UL ? mm->highest_vm_end : end, &clear_refs_walk_ops,
-				&cp);
+		walk_page_range(mm, start, end, &clear_refs_walk_ops, &cp);
 		if (type == CLEAR_REFS_SOFT_DIRTY) {
 			mmu_notifier_invalidate_range_end(&range);
 			flush_tlb_mm(mm);

